### PR TITLE
Snippet grammar fixes and minor formal improvements throughout the LSP spec

### PR DESCRIPTION
**Pull request by falko17**
While reading through the specification, I've noticed a few style issues as well as some shortcomings in the EBNF grammar for snippets. This PR should fix them. See the list below for details on what was changed.

## Detailed changes
* Most of the changes are various grammar, spelling, punctuation, formatting, or consistency improvements[^1].
* Some incorrect TypeScript syntax in the examples has been fixed.
* The EBNF grammar for snippets has been improved in the following ways:
  * Incorrect grouping of alternatives in some parts has been fixed.
  * Ambiguities present in the grammar have been fixed, the grammar should now be LL(1)-parseable.
  * The grammar has been made consistent with the description above it. Specifically, characters which have to be escaped per the textual description now also need to be escaped per the grammar rules[^2].
  * A description for the ':+', ':?', and ':-' modifiers has been added, as they seem to not be explained anywhere else in the spec.
     * I have used the VSCode source code as a basis for these.
  * The link has been updated to refer to the right EBNF specification, as the grammar uses the XML variant rather than the one described in the originally linked Wikipedia article.

> [!NOTE]
> These changes have only been applied to the upcoming 3.18 LSP specification. If they should be applied to the current version 3.17, I could port the changes over as well.

[^1]: Not all the changes necessarily fix errors, some just reword certain passages to sound more natural, which may be somewhat subjective.
[^2]: Technically, the text clarifying escaping rules could be removed, as the grammar already describes them now. Keeping the text for clarity's sake might still make sense, though.

### CODE REVIEW

**Comment by falko17** [comment]
@microsoft-github-policy-service agree

File: _specifications/lsp/3.18/language/codeAction.md [hunk 1]
```diff
@@ -400,8 +398,8 @@ export namespace CodeActionTriggerKind {
 	/**
 	 * Code actions were requested automatically.
 	 *
-	 * This typically happens when current selection in a file changes, but can
-	 * also be triggered when file content changes.
+	 * This typically happens when the current selection in a file changes,
+   * but can also be triggered when file content changes.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/codeAction.md]: Little nit: needs proper indentation.

File: _specifications/lsp/3.18/language/completion.md [hunk 2]
```diff
@@ -303,16 +303,16 @@ export interface CompletionContext {
 	triggerKind: CompletionTriggerKind;
 
 	/**
-	 * The trigger character (a single character) that has trigger code
-	 * complete. Is undefined if
+	 * The trigger character (a single character) that
+   * has triggered code complete. Is undefined if
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/completion.md]: Indentation.

File: _specifications/lsp/3.18/language/completion.md [hunk 3]
```diff
@@ -592,28 +591,28 @@ export interface CompletionItem {
...
 	 * for this item.
 	 *
 	 * The `insertText` is subject to interpretation by the client side.
-	 * Some tools might not take the string literally. For example
-	 * VS Code when code complete is requested in this example
-	 * `con<cursor position>` and a completion item with an `insertText` of
-	 * `console` is provided it will only insert `sole`. Therefore it is
+	 * Some tools might not take the string literally. For example,
+	 * when code complete is requested for `con<cursor position>`
+   * and a completion item with an `insertText` of `console` is provided,
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/completion.md]: Indentation

File: _specifications/lsp/3.18/language/completion.md [hunk 4]
```diff
@@ -641,11 +640,10 @@ export interface CompletionItem {
 
 	/**
 	 * An edit which is applied to a document when selecting this completion.
-	 * When an edit is provided the value of `insertText` is ignored.
+	 * When an edit is provided, the value of `insertText` is ignored.
 	 *
 	 * *Note:* The range of the edit must be a single line range and it must
-	 * contain the position at which completion has been requested. Despite this
-	 * limitation, your edit can write multiple lines.
+	 * contain the position at which completion has been requested.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/completion.md]: Why did you remove 

> Despite this limitation, your edit can write multiple lines.

I added this since clients weren't aware that this is possible

File: _specifications/lsp/3.18/language/completion.md [hunk 5]
```diff
@@ -656,12 +654,11 @@ export interface CompletionItem {
 	 * capability property.
 	 *
 	 * *Note 1:* The text edit's range as well as both ranges from an insert
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/completion.md:657]: Same with

> In both case the new text can have multiple lines.

File: _specifications/lsp/3.18/language/inlayHint.md [hunk 6]
```diff
@@ -140,8 +140,8 @@ export interface InlayHint {
 	 * hint (or its nearest variant) is now part of the document and the inlay
 	 * hint itself is now obsolete.
 	 *
-	 * Depending on the client capability `inlayHint.resolveSupport` clients
-	 * might resolve this property late using the resolve request.
+	 * Depending on the client capability `inlayHint.resolveSupport`,
+   * clients might resolve this property late using the resolve request.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/inlayHint.md]: Indentation

File: _specifications/lsp/3.18/language/selectionRange.md [hunk 7]
```diff
@@ -83,8 +83,8 @@ export interface SelectionRange {
 	 */
 	range: Range;
 	/**
-	 * The parent selection range containing this range. Therefore
-	 * `parent.range` must contain `this.range`.
+	 * The parent selection range containing this range.
+   * Therefore, `parent.range` must contain `this.range`.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/selectionRange.md]: Indentation

File: _specifications/lsp/3.18/language/signatureHelp.md [hunk 8]
```diff
@@ -203,44 +203,44 @@ _Response_:
...
- * active and only one active parameter.
+ * callable. There can be multiple signatures,
+ * but only one active one and only one active parameter.
  */
 export interface SignatureHelp {
 	/**
-	 * One or more signatures. If no signatures are available the signature help
-	 * request should return `null`.
+	 * One or more signatures. If no signatures are available,
+   * the signature help request should return `null`.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/signatureHelp.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 9]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
  * @since 3.17.0
  */
 export type NotebookDocumentFilter = {
-	/** The type of the enclosing notebook. */
+	/**
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md:171]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 10]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   * The type of the enclosing notebook.
+   */
 	notebookType: string;
 
-	/** A Uri [scheme](#Uri.scheme), like `file` or `untitled`. */
+	/**
+   * A Uri [scheme](#Uri.scheme), like `file` or `untitled`.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 11]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	scheme?: string;
 
-	/** A glob pattern. */
+	/**
+   * A glob pattern.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 12]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	pattern?: string;
 } | {
-	/** The type of the enclosing notebook. */
+	/**
+   * The type of the enclosing notebook.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 13]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	notebookType?: string;
 
-	/** A Uri [scheme](#Uri.scheme), like `file` or `untitled`.*/
+	/**
+   * A Uri [scheme](#Uri.scheme), like `file` or `untitled`.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 14]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	scheme: string;
 
-	/** A glob pattern. */
+	/**
+   * A glob pattern.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 15]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	pattern?: string;
 } | {
-	/** The type of the enclosing notebook. */
+	/**
+   * The type of the enclosing notebook.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 16]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	notebookType?: string;
 
-	/** A Uri [scheme](#Uri.scheme), like `file` or `untitled`. */
+	/**
+   * A Uri [scheme](#Uri.scheme), like `file` or `untitled`.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/notebookDocument/notebook.md [hunk 17]
```diff
@@ -168,51 +168,69 @@ export interface NotebookCellTextDocumentFilter {
...
+   */
 	scheme?: string;
 
-	/** A glob pattern. */
+	/**
+   * A glob pattern.
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/notebookDocument/notebook.md]: Indentation

File: _specifications/lsp/3.18/language/completion.md [hunk 18]
```diff
@@ -814,28 +811,36 @@ ${TM_FILENAME/(.*)\..+$/$1/}
...
                 | '${' int ':-' else '}' | '${' int ':' else '}'
 regex       ::= Regular Expression value (ctor-string)
 options     ::= Regular Expression option (ctor-options)
 var         ::= [_a-zA-Z] [_a-zA-Z0-9]*
 int         ::= [0-9]+
-text        ::= .*
-if			::= text
-else		::= text
+text        ::= ([^$}\] | '\$' | '\}' | '\\')*
+choicetext  ::= ([^,|\] | '\,' | '\|' | '\\')*
```
Comment (dbaeumer) [inline _specifications/lsp/3.18/language/completion.md:841]: IMO we need to make choicetext as follows:

`([^$},|\] | '\$' | '\}' |'\,' | '\|' | '\\')*`

Any reason why you didn't

**Comment by dbaeumer** [comment]
@falko17 thanks you so much for the great PR. Highly appreciated. I have several comments about indentation. If you are too busy to fix them I am happy to do so. I have one additional comment about the changes you did to the snippet syntax I would like to get your comment on.

**Review by dbaeumer — CHANGES_REQUESTED** [review]
See all my inline comments and my final one.

File: _specifications/lsp/3.18/language/completion.md (hunk 4 above)
Comment (falko17) [reply _specifications/lsp/3.18/language/completion.md]: I probably thought that part was redundant as it's implicitly true given the definition of ranges, but if there already was confusion about this from clients, I'll add it back—my bad!

File: _specifications/lsp/3.18/language/completion.md (hunk 5 above)
Comment (falko17) [reply _specifications/lsp/3.18/language/completion.md:657]: Yes, I'll add it back.

File: _specifications/lsp/3.18/language/completion.md (hunk 18 above)
Comment (falko17) [reply _specifications/lsp/3.18/language/completion.md:841]: The reason why `$` and `}` don't need to be escaped within a choice construct is that it cannot occur in the `choice` in as part of any other construct, so it is uniquely parseable. Only a `,`—indicating the next `choicetext`—and a `|}`—indicating the end of the `choice` construct—can occur here as tokens with semantics attached. Anything else (including `$` and `}`) can be ignored by the parser, and thus doesn't need to be escaped.

For example, `${1|close},$variable|}` represents a choice construct with choices `close}` and `$variable`. These two choices don't have further meaning and are interpreted purely as strings. Contrast this with e.g. `${1|one,\,,two}`, which is a choice construct with choices `one`, `,`, and `two`. If the comma here were not escaped, it would be unclear whether the choice consists of two empty elements or a comma, hence the escape is necessary.

(Formally, $FIRST(\texttt{choicetext}) \cap FOLLOW(\texttt{choicetext}) = \emptyset$ must be true here[^1]. As $FOLLOW(\texttt{choicetext}) = \\{'\texttt{|}',\ '\texttt{,}'\\}$, it is only these two characters that need to be escaped, so that they are not part of $FIRST(\texttt{choicetext})$.)

[^1]: FIRST/FIRST conflicts can occur too, but are only relevant insofar that the backslash needs to be escaped.

**Comment by falko17** [comment]
Thank you for the detailed review, @dbaeumer! I've addressed your comments, please tell me if there is anything else that should be done. Also, sorry about all the indentation problems, my editor displayed tabs as 2 spaces, which is why I didn't catch these. Should be all fixed now.

File: _specifications/lsp/3.18/language/completion.md (hunk 18 above)
Comment (dbaeumer) [reply _specifications/lsp/3.18/language/completion.md:841]: @falko17 thanks for the great explanation!

**Review by dbaeumer — APPROVED** [review]

**Review by lszomoru — APPROVED** [review]
